### PR TITLE
WIP [DO NOT MERGE]: Topology-aware Comm_split_type

### PR DESCRIPTION
**Pull request by pavanbalaji**
Allow for hwloc integration and using it to create communicators based on on-node topology information.

### CODE REVIEW

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/all

**Comment by pavanbalaji** [comment]
test:jenkins/ch4/all

**Comment by KavithaTipturMadhu** [comment]
test:jenkins/ch3/all
test:jenkins/ch4/all

File: src/hwloc/config/hwloc.m4 [hunk 1]
```diff
@@ -1179,15 +1176,21 @@ EOF])
 
     AS_IF([test "$hwloc_mode" = "embedded"],
           [HWLOC_EMBEDDED_CFLAGS=$HWLOC_CFLAGS
+           HWLOC_EMBEDDED_VISIBILITY_CFLAGS=$HWLOC_VISIBILITY_CFLAGS
```
Comment (raffenet) [inline src/hwloc/config/hwloc.m4]: What are embedded visibility flags? Embedding in my mind means static linking. Visibility only applies to shared libraries.

Comment (raffenet) [reply src/hwloc/config/hwloc.m4]: On second thought, this sounds familiar to an issue I ran into with MPL and visibility where I had to do something similar.

Comment (pavanbalaji) [reply src/hwloc/config/hwloc.m4]: The visibility flags are included in the MPICH build, and will cause MPICH to not build correctly in some cases.  The above patch allows us to distinguish those cases and disable them in those cases.

File: src/mpi/comm/comm_split_type.c [hunk 2]
```diff
@@ -73,6 +70,142 @@ int MPIR_Comm_split_type_node(MPIR_Comm * comm_ptr, int key, MPIR_Comm ** newcom
...
+    hwloc_obj_type_t hwloc_obj_type;
+    int depth;
+};
+
+static struct shmem_info_table shmem_info[] = {
+    {"socket", HWLOC_OBJ_PACKAGE, -1},
+    {"package", HWLOC_OBJ_PACKAGE, -1},
+    {"numanode", HWLOC_OBJ_NUMANODE, -1},
+    {"core", HWLOC_OBJ_CORE, -1},
+    {"hwthread", HWLOC_OBJ_PU, -1},
```
Comment (pavanbalaji) [inline src/mpi/comm/comm_split_type.c]: There should be an alias for "pu" to point to `HWLOC_OBJ_PU` as well.

Comment (pavanbalaji) [reply src/mpi/comm/comm_split_type.c]: We should probably add equivalents for other objects in hwloc too, e.g., `MACHINE`.

File: src/mpi/comm/comm_split_type.c [hunk 3]
```diff
@@ -73,6 +70,142 @@ int MPIR_Comm_split_type_node(MPIR_Comm * comm_ptr, int key, MPIR_Comm ** newcom
...
+    {"l1cache", HWLOC_OBJ_CACHE, 1},
+    {"l2cache", HWLOC_OBJ_CACHE, 2},
+    {"l3cache", HWLOC_OBJ_CACHE, 3},
```
Comment (pavanbalaji) [inline src/mpi/comm/comm_split_type.c]: Should these be split into data, instruction and unified caches, and default to data/unified caches if unspecified?

File: src/mpi/comm/comm_split_type.c [hunk 4]
```diff
@@ -73,6 +70,142 @@ int MPIR_Comm_split_type_node(MPIR_Comm * comm_ptr, int key, MPIR_Comm ** newcom
...
+};
+
+static int comm_split_type_hwloc(const char *hardware_filter, int node_id, MPIR_Comm * comm_ptr,
+                                 MPIR_Comm ** newcomm_ptr, int key)
+{
+    int color = node_id;
+    int i;
+    int mpi_errno = MPI_SUCCESS;
+
+    if (!MPIR_Process.bindset_is_valid)
```
Comment (pavanbalaji) [inline src/mpi/comm/comm_split_type.c]: In what case will the bindset not be valid?

File: src/mpi/comm/comm_split_type.c [hunk 5]
```diff
@@ -73,6 +70,142 @@ int MPIR_Comm_split_type_node(MPIR_Comm * comm_ptr, int key, MPIR_Comm ** newcom
...
+            hwloc_obj_type_t query_obj_type = shmem_info[i].hwloc_obj_type;
+            hwloc_obj_t obj_containing_cpuset;
+
+            /* we deal with cache objects separately since they
+             * require an additional depth query */
+            if (query_obj_type == HWLOC_OBJ_CACHE) {
+                obj_containing_cpuset =
+                    hwloc_get_cache_covering_cpuset(MPIR_Process.topology, MPIR_Process.bindset);
+                if (obj_containing_cpuset == NULL)
+                    goto do_split;
```
Comment (pavanbalaji) [inline src/mpi/comm/comm_split_type.c]: In cases where the user asked us to split a certain way and we could not find the right object to do that, we should at least throw a warning saying so.

**Pull request closed without merging**